src/skagent/model_analyzer.py

```diff
@@ -0,0 +1,274 @@
+
+    def analyze(self):
+        """Run the full analysis pipeline."""
+        self._collect_nodes()
+        self._collect_dependencies()
+
+        # Treat any referenced-but-unknown symbol as exogenous state
+        for tgt, deps in list(self._raw_deps.items()):
+            for src in deps:
+                if src not in self.node_meta:
+                    self.node_meta[src] = {
+                        "kind": "state",
+                        "agent": "global",
+                        "plate": None,
+                        "observed": False,
+                    }
+
+        self._identify_time_dependencies()
+        self._assemble_edges()
+        self._collect_formulas()
+        self._collect_plates()
+        return self
+
+    def _collect_nodes(self):
+        """Classify every variable and record its metadata."""
+        # 1) all LHS names in dynamics & reward
+        lhs = set()
+        for blk in self._blocks:
+            lhs |= set(blk.dynamics.keys())
+            lhs |= set(blk.reward.keys())
+
+        # 2) shocks
+        for blk in self._blocks:
+            for var, shock_def in blk.shocks.items():
+                agent = getattr(shock_def, "agent", "global") or "global"
+                self.node_meta[var] = {
+                    "kind": "shock",
+                    "agent": agent,
+                    "plate": agent if agent != "global" else None,
+                    "observed": False,
+                }
+
+        # 3) dynamics
+        for blk in self._blocks:
+            for var, rule in blk.dynamics.items():
+                kind = "control" if isinstance(rule, Control) else "state"
+                agent = getattr(rule, "agent", "global") or "global"
+                self.node_meta[var] = {
+                    "kind": kind,
+                    "agent": agent,
+                    "plate": agent if agent != "global" else None,
+                    "observed": (kind in ("control", "reward"))
+                    or (var in self.observables),
+                }
+
+        # 4) reward
+        for blk in self._blocks:
+            for var, rd in blk.reward.items():
+                agent = (
+                    rd
+                    if isinstance(rd, str)
+                    else getattr(rd, "agent", "global") or "global"
+                )
+                self.node_meta[var] = {
+                    "kind": "reward",
+                    "agent": agent,
+                    "plate": agent if agent != "global" else None,
+                    "observed": True,
+                }
+
+        # 5) params: only keys in calibration not appearing on any LHS
+        for p in set(self.calibration) - lhs:
+            self.node_meta.setdefault(
+                p,
+                {"kind": "param", "agent": "global", "plate": None, "observed": False},
+            )
+
+    def _rhs_symbols(self, rule):
+        """Extract variable names from RHS rule (callable or string)."""
```

Comment:
This is going to miss the following cases:
 - when the RHS is a Control with an information set
 - when the RHS is a Distribution that depends on variables that are in the calibration